### ISSUE TITLE
The android-ndk is planning to drop i686/x86 support

### ISSUE THREAD

**Issue by Grimler91** [termux/termux-packages#12954]
i686/x86 support will be dropped from android-ndk, possibly already from ndk-r26: https://github.com/android/ndk/issues/1772

If we are to keep updating i686 we would hence have to stick on an older ndk version.

How many users do we have on that arch? (I will make some statistics of downloads from fosshost/grimler.se later)

Looking at some old issues we have had bug reports from (list not conclusive): @snogglethorpe, @rasmusolle, @amsitlab, @qntmpkts, @petjal on i686. Are any of you guys still using termux on these devices (i.e. would dropping i686 be problematic for you)?

**Comment by buttaface** [comment]
Probably almost nobody, it never had much support.

**Comment by Grimler91** [comment]
During the last 14 days, 1.8 % of the downloads on fosshost and 1.1 % on grimler.se have been i686 debs. x86_64 has even less downloads, 1.2 % on fosshost and 0.7 % on grimler.se. 

On grimler.se the downloads were done by ~850 different ip addresses, so while a small percentage it still many thousands of users on all the mirrors

**Comment by xtkoba** [comment]
If we are going to keep maintaining i686 builds, then can't we use a different NDK version for i686 from the others, just as we did lately for the sake of `-fno-integrated-as`.

**Comment by Grimler91** [comment]
Yeah, that would work but potentially be a bit messy when we have to workaround issues/quirks for different ndk versions

**Comment by truboxl** [comment]
A hypothetical question is how to maintain TERMUX_NDK_VERSION dependent packages if future NDK breaks the last supported x86 NDK?

**Comment by xtkoba** [comment]
A possible way would be to create e.g. `libc++-i686` for `i686` only that "provides" `libc++` and exclude `i686` from `libc++`. Very ugly and should be avoided though.

**Comment by buttaface** [comment]
I'd rather just drop i686 updates, but keep the old packages up for people who want the last versions, along with a warning that they're not updated.

**Comment by xtkoba** [comment]
I have no strong opinion about whether to drop support for i686 or not itself, but I am opposed to sticking to an old version of NDK for every architecture, because in that manner we cannot support new architectures that will be supported by future versions of NDK, such as riscv64 (I don't know if it will be actually supported but there are activities in the Bionic codebase.)

**Comment by buttaface** [comment]
> I am opposed to sticking to an old version of NDK for every architecture, because in that manner we cannot support new architectures that will be supported by future versions of NDK

I don't think that will be an issue now that we support building with multiple NDK versions. We could simply change the build logic so that i686 is only built with the last NDK that supported it, while other arches would be built with the latest NDK.

**Comment by buttaface** [comment]
Btw, [32-bit support is starting to get dropped altogether on the latest devices](https://arstechnica.com/gadgets/2022/10/the-pixel-7-tests-out-64-bit-only-android-cant-run-32-bit-apps/), with most of the new ARMv9 cores not supporting AArch32 anymore.

We should also consider dropping 32-bit at some point: what do the armv7 download percentages look like today, @Grimler91?

The good news is that we will soon have [a new architecture to support](https://arstechnica.com/gadgets/2023/01/google-announces-official-android-support-for-risc-v/).

**Comment by xtkoba** [comment]
As a moderator, I see not a few number of `termux-info` output for arm (32-bit) architecture even with rather new devices, unfortunately. Search the issues for "Packages CPU architecture: arm" keyword. I don't think we can drop arm (32-bit) support within a handful of years. But definitly we can after 15 years from now, due to Y2038.

**Comment by xtkoba** [comment]
It would be more plausible to drop support for API less than 26 or 28 or so, than to drop arm (32-bit) support. Sorry for being off-topic.

**Comment by buttaface** [comment]
Btw, it looks like [the upcoming Android 14 disallows installing apps targeting API 23 or older](https://arstechnica.com/gadgets/2023/02/android-14-preview-1-is-out-will-officially-ban-installation-of-old-apps/), which doesn't affect us but will within 5 years, as they appear to plan to keep raising that cutoff target API in the coming years.

If we cannot target Android 10 or later, it looks like we have about 5 years before Termux will not install on Android, unless you use the special adb flag the article mentions.

**Comment by truboxl** [comment]
A bit off topic: we actually been building i686 programs using lower than Google's minimum requirement flags for a long time.

For reference the current Google's suggested optimization: https://developer.android.com/ndk/guides/abis#x86
```
-march=i686 -mtune=intel -mssse3 -mfpmath=sse -m32
```

I am not sure when did Google silently bump up the requirement for x86. Checking AOSP history will take some time.

Our current build flags for i686:

https://github.com/termux/termux-packages/blob/0df483d6b91bffa4c17d99bb60339c53e31dd239/scripts/build/toolchain/termux_setup_toolchain_25c.sh#L58

dates all the way back to the initial commit of this repo
https://github.com/termux/termux-packages/blob/59f0d218a6ff34c80cf898f6d7ac62555ba8eb11/build-package.sh#L101

So cutting off i686 support really means we are cutting support for "wider than usual" range of x86 devices if there even is one

Obviously this is not a proposal to bump up the requirement one last time to match Google's suggestion for a deprecating platform anyway...
